### ISSUE TITLE
Killing zvol_id during zpool import

### ISSUE THREAD

**Issue by arturpzol** [zfsonlinux/zfs#3659]
Hello,

I am using ZOL 0.6.4.2 in cluster environment with shared storage. Debian Jessie with kernel 3.10.

Two nodes and each node has own pool with some zvols which are shared via iSCSI and some datasets which are shared via NFS. When I shutdown e.g. node A, second node B takeovers pool from node A. After import pool I noticed that some symlinks for zvols are missing.

```
...
/dev/Pool-1
/dev/Pool-1/vol70-part1
/dev/Pool-1/vol70
/dev/Pool-1/vol51-part1
/dev/Pool-1/vol51
/dev/Pool-1/vol41-part1
/dev/Pool-1/vol60-part1
/dev/Pool-1/vol41
/dev/Pool-1/vol60
/dev/Pool-1/vol71-part1 -> /dev/zd176p1
/dev/Pool-1/vol50-part1
/dev/Pool-1/vol40-part1
/dev/Pool-1/vol61-part1
/dev/zd272
/dev/zd256p1
/dev/zd256
/dev/zd240p1
/dev/zd240
/dev/zd224p1
/dev/zd224
/dev/zd208p1
/dev/zd208
/dev/zd192p1
/dev/zd192
/dev/zd176p1
/dev/zd176
/dev/zd160p1
/dev/zd160
/dev/zd144p1
/dev/zd144
...
```
In above case symlink /dev/Pool-1/vol71 is not available but devices /dev/zd176p1
/dev/zd176 exists. 

I enabled debug for udev and logs showed such error:

```timeout: killing '/lib/udev/zvol_id /dev/zd176' [25722]```

Have you experienced similar issue? 

When I changed two entries in 60-persistent-storage.rules file to:

```
# probe filesystem metadata of disks
KERNEL!="sr*|zd*", IMPORT{builtin}="blkid"'

# watch metadata changes by tools closing the device after writing
KERNEL!="sr*|zd*", OPTIONS+="watch"
```

problem solved.

**Comment by behlendorf** [comment]
@arturpzol thanks for the workaround.  To be clear, your patch resolves the issue by disabling blkid so it doesn't automatically run on zvols.  That's probably not what we want the default behavior to be but I can see it definitely being useful when you have a lot of volumes.

I was thinking another solution to this issue might be to have ZFS automatically prefetch the first and last blocks for the volume when it creates the minor.  This should significantly speed up `blkid` which scans these areas of the volume by turning them in the ARC cache hits.

I've opened #3691 with a patch which implements this, if you get a chance to test it out that would be appreciated.  I've added the `zvol_prefetch_bytes` module option to control how much prefetching happens so we can seltle on a good default value.  For the moment I've made an educated guess of 128K.

**Comment by arturpzol** [comment]
@behlendorf I have performed a few tests but unfortunately killing zvol_id still occur. Should I change zvol_prefetch_bytes value to other?

**Comment by behlendorf** [comment]
@arturpzol can you try increasing `zvol_prefetch_bytes`, that was just a best initial guess.  Can you tell if it helped any?

**Comment by Bronek** [comment]
@behlendorf any chance of this being included in 0.6.5 ? I have similar performance problem (although to smaller degree) on my zvols

**Comment by behlendorf** [comment]
@Bronek unfortunately the proposed `60-persistent-storage.rule` patch can't be included in ZFS because it modified files owned by a different package.  You can definitely make the same modification locally to resolve the issue.  As for the 9965059 patch this I have included in 0.6.5 and it should help somewhat, so it may be enough for your situation.

I'm leaving this issue as a reminded to revisit this issue so we can make additional improvements.

**Comment by Bronek** [comment]
@behlendorf thanks, yes your commit will be sufficient.

**Comment by behlendorf** [comment]
Closing.  This has been addressed in 0.6.5.6 by making zvol minor creation asynchronous to `zpool import` and by prefetching the first/last blocks of a zvol for blkid.   The `60-persistent-storage.rule` isn't shipped by ZoL so this isn't something we can change.
